collector/aws/elasticsearch.go

```diff
@@ -0,0 +1,290 @@
+
+	return &ElasticSearchManager{
+		collector:          collector,
+		client:             client,
+		cloudWatchCLient:   cloudWatchCLient,
+		metrics:            metrics,
+		pricingClient:      pricing,
+		stsClient:          sts,
+		region:             region,
+		namespace:          "AWS/ES",
+		servicePricingCode: "AmazonES",
+		Name:               fmt.Sprintf("%s_elasticsearch", ResourcePrefix),
+	}
+}
+
+// Detect checks with elasticsearch cluster is underutilized
+func (esm *ElasticSearchManager) Detect() ([]DetectedElasticSearch, error) {
+
+	log.WithFields(log.Fields{
+		"region":   esm.region,
+		"resource": "elasticsearch",
+	}).Info("analyzing resource")
+
+	esm.collector.UpdateServiceStatus(collector.EventCollector{
+		ResourceName: esm.Name,
+		Data: collector.EventStatusData{
+			Status: collector.EventFetch,
+		},
+	})
+
+	detectedElasticSearchClusters := []DetectedElasticSearch{}
+
+	clusters, err := esm.DescribeClusters()
+	if err != nil {
+
+		esm.collector.UpdateServiceStatus(collector.EventCollector{
+			ResourceName: esm.Name,
+			Data: collector.EventStatusData{
+				Status:       collector.EventError,
+				ErrorMessage: err.Error(),
+			},
+		})
+
+		return detectedElasticSearchClusters, err
+	}
+
+	now := time.Now()
+
+	for _, cluster := range clusters {
+		log.WithField("cluster_arn", *cluster.ARN).Debug("checking elasticsearch cluster")
+
+		instancePrice, err := esm.pricingClient.GetPrice(esm.GetPricingFilterInput([]*pricing.Filter{
+			{
+				Type:  awsClient.String("TERM_MATCH"),
+				Field: awsClient.String("instanceType"),
+				Value: awsClient.String(*cluster.ElasticsearchClusterConfig.InstanceType),
+			},
+		}), "", esm.region)
+		if err != nil {
+			log.WithError(err).Error("Could not get instance price")
+			continue
+		}
+
+		var EBSPrice float64
+		var hourlyEBSVolumePrice float64
+		if *cluster.EBSOptions.EBSEnabled {
+			if storageMedia, found := elasticSearchVolumeType[*cluster.EBSOptions.VolumeType]; found {
+				EBSPrice, err = esm.pricingClient.GetPrice(esm.GetPricingFilterInput([]*pricing.Filter{
+					{
+						Type:  awsClient.String("TERM_MATCH"),
+						Field: awsClient.String("storageMedia"),
+						Value: awsClient.String(storageMedia),
+					},
+				}), "", esm.region)
+				if err != nil {
+					log.WithError(err).Error("Could not get ebs price")
+					continue
+				}
+				hourlyEBSVolumePrice = (EBSPrice * float64(*cluster.EBSOptions.VolumeSize)) / collector.TotalMonthHours
+			}
+		}
+
+		log.WithFields(log.Fields{
+			"instance_hour_price": instancePrice,
+			"ebs_hour_price":      hourlyEBSVolumePrice,
+			"region":              esm.region}).Debug("Found the following price list")
+
+		callerIdentityOutput, err := esm.stsClient.client.GetCallerIdentity(&sts.GetCallerIdentityInput{})
```

Comment:
why are you calling this function on loop? 
the account id can be changed? 

and for my perspective, you don't need the sts client. you only need to get the account id, so why not calling this function once in `run.go` ?

Reply:
moveed to run.go